### PR TITLE
Change linkedin flow

### PR DESCRIPTION
**Pull request by nwjbrandon**

### CODE REVIEW

File: src/api.py [hunk 1]
```diff
@@ -1,15 +1,42 @@
...
 
 blueprint = Blueprint("root", version="v1")
 
 
-def auth_required(func):
-    return (inject_user(blueprint))(protected(blueprint)(func))
+def auth_required(f):
+    @wraps(f)
+    async def decorated_function(request, *args, **kwargs):
+        def get_token(header):
```
Comment (donjar) [inline src/api.py]: I would prefer manipulating the `request.headers['Authorization']` directly instead of creating another fn here

File: src/api.py [hunk 2]
```diff
@@ -1,15 +1,42 @@
...
+        token = get_token(request.headers["Authorization"])
+        linkedin_user = request.app.social_login.get_linkedin_user(token=token)
+        user = request.app.user_service.get_user_by_linkedin_id(
+            user_id=linkedin_user.get("user_id")
+        )
+        if user is not None:
+            response = await f(request, user, *args, **kwargs)
+            return response
+        else:
+            return json({"status": "not_authorized"}, 403)
```
Comment (donjar) [inline src/api.py]: Raise an exception in `src/exceptions.py` instead of returning something here. (Also please use a 404 one)

File: src/api.py [hunk 3]
```diff
@@ -1,15 +1,42 @@
...
+
+    return decorated_function
+
+
+@blueprint.get("/auth/me")
+@auth_required
+async def test(request, user):
```
Comment (donjar) [inline src/api.py]: Change the function name

File: src/api.py [hunk 4]
```diff
@@ -26,6 +53,7 @@ def auth_required(func):
 
 @expects_json_object
 async def user_login(request):
+    print(request)
```
Comment (donjar) [inline src/api.py]: Remove this

File: src/database.py [hunk 5]
```diff
@@ -58,11 +58,13 @@ class User(Base):
     __tablename__ = "users"
 
     email = Column(String, nullable=False, unique=True)
-    hashed_password = Column(String, nullable=False)
+    provider = Column(String, nullable=False)
     full_name = Column(String, nullable=False)
+    display_image = Column(String)
```
Comment (donjar) [inline src/database.py]: Can we change this to `display_image_url` to make it clearer

File: src/services.py [hunk 6]
```diff
@@ -135,18 +126,16 @@ def get_user(self, id):
...
             user_dict = user.asdict()
-        user_dict.pop("hashed_password")
         return user_dict
 
-    def get_user_by_email(self, email):
+    def get_user_by_linkedin_id(self, user_id):
         with session_scope() as session:
-            user = session.query(User).filter_by(email=email).one_or_none()
+            user = session.query(User).filter_by(user_id=user_id).one_or_none()
             if user is None:
```
Comment (donjar) [inline src/services.py:132]: Double consecutive `if user is None` here

File: src/services.py [hunk 7]
```diff
@@ -48,22 +49,27 @@ def __init__(self, config, hasher=argon2):
...
+                    full_name=full_name,
+                    display_image=display_image,
+                    provider="linkedin",
+                    can_buy=False,
+                    can_sell=False,
+                    user_id=user_id,
+                )
+                session.add(user)
+            else:
+                user.display_image = display_image
```
Comment (donjar) [inline src/services.py]: If the user exists, why still change these stuff?

File: src/services.py [hunk 8]
```diff
@@ -694,15 +684,23 @@ def get_auth_url(self, socket_id):
...
         token = self.get_token(code=code)
-        full_name = self.get_user_profile(token=token)
-        email = self.get_user_email(token=token)
-        user = UserService(self.config).get_user_by_email(email=email)
-        user = UserService(self.config).activate_buy_privileges(user_id=user.get("id"))
-        user["created_at"] = user.get("created_at").timestamp()
-        user["updated_at"] = user.get("updated_at").timestamp()
-        await self.sio.emit("provider", user, namespace="/v1/", room=socket_id)
+        user = self.get_linkedin_user(token)
+        user = self.UserService(self.config).create(
```
Comment (donjar) [inline src/services.py]: No need the `user = `?

Comment (donjar) [reply src/services.py]: Also I think you can just do `.create(**user)` or something

Comment (donjar) [reply src/services.py]: And let's rename this to be `create_if_not_exists` or something

File: src/services.py [hunk 9]
```diff
@@ -677,9 +666,10 @@ def get_other_party_details(self, chat_room_id, user_id):
 
 
 class SocialLogin:
-    def __init__(self, config, sio):
+    def __init__(self, config, sio, UserService=UserService):
         self.config = config
         self.sio = sio
+        self.UserService = UserService
```
Comment (donjar) [inline src/services.py]: Let's get rid of this `self.UserService` etc injection stuff. I felt that it's not very useful

File: src/services.py [hunk 10]
```diff
@@ -677,9 +666,10 @@ def get_other_party_details(self, chat_room_id, user_id):
 
 
 class SocialLogin:
```
Comment (donjar) [inline src/services.py]: Shall we name it `LinkedinLogin` to make it clearer? Then if we support more providers we can have different names.

**Review by donjar — COMMENTED** [review]
Can you remove the JWT library as well? `poetry remove sanic-jwt` or something.

File: src/services.py (hunk 7 above)
Comment (nwjbrandon) [reply src/services.py]: Because if the user updates their email, or change user profile, then when they login, these fields get updated.

Is that alright?

File: src/app.py [hunk 11]
```diff
@@ -60,12 +58,12 @@ def sentry_before_send(event, hint):
 app.banned_pair_service = BannedPairService(app.config)
 app.chat_room_service = ChatRoomService(app.config)
 app.chat_service = ChatService(app.config)
-app.social_login = SocialLogin(app.config, sio)
+app.linkedin_login = LinkedInLogin(app.config, sio)
 
 initialize_cors(app)
 
 
-class AcquityJwtResponses(Responses):
```
Comment (donjar) [inline src/app.py:70]: Why is this changed?

**Review by donjar — COMMENTED** [review]
Can you remove the JWT related config in `src/config.py`?
